gpcontrib/arenadata_toolkit/src/arenadata_toolkit_worker.c

```diff
@@ -0,0 +1,450 @@
+{
+	Relation	rel;
+	SysScanDesc scan;
+	HeapTuple	tup;
+	List	   *dbids = NIL;
+	tracked_db_t *trackedDb;
+
+	rel = heap_open(DbRoleSettingRelationId, RowExclusiveLock);
+	scan = systable_beginscan(rel, InvalidOid, false, NULL, 0, NULL);
+	while (HeapTupleIsValid(tup = systable_getnext(scan)))
+	{
+		bool		isnull;
+		Datum		str_datum;
+		Datum		oid_datum;
+		ArrayType  *a;
+
+		str_datum = heap_getattr(tup, Anum_pg_db_role_setting_setconfig,
+								 RelationGetDescr(rel), &isnull);
+		if (isnull)
+			continue;
+
+		oid_datum = heap_getattr(tup, Anum_pg_db_role_setting_setrole,
+								 RelationGetDescr(rel), &isnull);
+		if (DatumGetObjectId(oid_datum) != InvalidOid)
+			continue;
+
+		oid_datum = heap_getattr(tup, Anum_pg_db_role_setting_setdatabase,
+								 RelationGetDescr(rel), &isnull);
+		if (DatumGetObjectId(oid_datum) == InvalidOid)
+			continue;
+
+		a = DatumGetArrayTypeP(str_datum);
+
+		if (is_db_tracked(a))
+		{
+			trackedDb = (tracked_db_t *) palloc0(sizeof(tracked_db_t));
+
+			trackedDb->dbid = DatumGetObjectId(oid_datum);
+			trackedDb->get_full_snapshot_on_recovery = full_snapshot_on_recovery(a);
+			dbids = lappend(dbids, trackedDb);
+		}
+	}
+
+	systable_endscan(scan);
+	heap_close(rel, RowExclusiveLock);
+
+	if (dbids != NIL)
+	{
+		ListCell   *cell;
+
+		if (dispatch_to_master)
+			dispatch_register_to_master(dbids);
+
+		dispatch_register_to_segments(dbids, uninitialized_segments);
+
+		foreach(cell, dbids)
+		{
+			pfree(lfirst(cell));
+		}
+
+		list_free(dbids);
+	}
+
+	LWLockAcquire(tf_shared_state->bloom_set.lock, LW_EXCLUSIVE);
+	tf_shared_state->bgworker_ready = true;
+	LWLockRelease(tf_shared_state->bloom_set.lock);
+}
+
+/* scan pg_db_role_setting, find all databases, bind blooms if necessary */
+static void
+arenadata_toolkit_worker(Datum main_arg)
+{
+	elog(LOG, "[arenadata toolkit] Starting background worker");
+
+	bool		master_initialized = false;
+
+	pqsignal(SIGHUP, tracking_sighup);
+	pqsignal(SIGTERM, tracking_sigterm);
+	pqsignal(SIGUSR1, tracking_sigusr1);
+
+	BackgroundWorkerUnblockSignals();
+
+	BackgroundWorkerInitializeConnection(DB_FOR_COMMON_ACCESS, NULL);
+
+	while (!got_sigterm)
+	{
+		int			rc;
+		List	   *uninitialized_segments = NIL;
+
+		CHECK_FOR_INTERRUPTS();
+
+		StartTransactionCommand();
+
+		if (extension_created())
+		{
+			elog(LOG, "[arenadata toolkit] Getting uninitialized segments");
+			uninitialized_segments = get_uninitialized_segments(uninitialized_segments);
+
+			if (!master_initialized || list_length(uninitialized_segments) > 0)
+			{
+				elog(LOG, "Dispatching register to segments");
+				dispatch_register(!master_initialized, uninitialized_segments);
+				list_free(uninitialized_segments);
+				uninitialized_segments = NIL;
+				master_initialized = true;
+			}
+		}
+		CommitTransactionCommand();
+
+		rc = WaitLatch(&MyProc->procLatch, WL_LATCH_SET | WL_TIMEOUT | WL_POSTMASTER_DEATH,
+					   tracking_worker_naptime_sec * 1000);
```

Comment:
If we send signals very often, we will exit the `WaitLatch` function more often than `tracking_worker_naptime_sec * 1000`. Here we need to use logic with `INSTR_TIME_SET_CURRENT` and `INSTR_TIME_SUBTRACT`, like [this](https://github.com/RekGRpth/pg_quota/blob/da335c8bdd0a421ec1508a0a7002a0b5b1147f94/pg_quota.c#L428-L438).

Reply:
I failed to understand, how the the time [difference](https://github.com/RekGRpth/pg_quota/blob/da335c8bdd0a421ec1508a0a7002a0b5b1147f94/pg_quota.c#L438) is used?

Reply:
The `pg_quota_timeout` function is called no more often than once every `pg_quota.timeout` milliseconds, even though the `WaitLatch` function may complete much more frequently due to signals.

Reply:
You should move the logic that is executed periodically into a separate function and call it only if at least tracking_worker_naptime_sec seconds have passed since the previous call.